### ISSUE TITLE
Lorax server hangs on warming up the server when total_tokens is larger than 4096 with LLama2-7b-hf on a l4 GPU

### ISSUE THREAD

**Issue by martinbomio** [predibase/lorax#334]
### System Info

These are the params I am using to start the lorax server (in GKE):
```
        - '--model-id=meta-llama/Llama-2-7b-hf'
        - '--revision=8cca527612d856d7d32bd94f8103728d614eb852'
        - '--source=hub'
        # greater than the max input tokens we will use in the benchmarks
        - '--max-input-length=1048'
        # greater than the sum of max output tokens + max input tokens
        # for any benchmark. The max we have is 512 input tokens + 3584 output tokens
        - '--max-total-tokens=4096'
        - '--max-batch-prefill-tokens=2048'
        # run on a single GPU for now
        - '--num-shard=1'
        - '--cuda-memory-fraction=0.9'
        - '--port=8080'
 ```

Running it on a single L4 GPU.
Using `ghcr.io/predibase/lorax:latest` image

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [ ] An officially supported command
- [X] My own modifications

### Reproduction

1. Use the following manifest and deploy it to kubernetes:
```
apiVersion: v1
kind: Pod
metadata:
  name: lorax
  namespace: namespace
  labels:
    owner: owner
    app: lorax
spec:
  containers:
    - name: lorax
      image: ghcr.io/predibase/lorax:latest
      env:
        - name: HUGGING_FACE_HUB_TOKEN
          value: ...
      args:
        - '--model-id=meta-llama/Llama-2-7b-hf'
        - '--revision=8cca527612d856d7d32bd94f8103728d614eb852'
        - '--source=hub'
        - '--max-input-length=1048'
        - '--max-total-tokens=4096'
        - '--max-batch-prefill-tokens=2048'
        - '--num-shard=1'
        - '--cuda-memory-fraction=0.9'
        - '--port=8080'
      ports:
        - name: http
          containerPort: 8080
          protocol: TCP
      readinessProbe:
        httpGet:
          path: /health
          port: http
        initialDelaySeconds: 15
        periodSeconds: 5
        timeoutSeconds: 10
      resources:
        requests:
          cpu: 6
          memory: 16G
        limits:
          cpu: 6
          memory: 16G
          nvidia.com/gpu: 1
      volumeMounts:
        - name: shm
          mountPath: /dev/shm
  nodeSelector:
    <select the l4 node pool>
  volumes:
    - name: shm
      emptyDir:
        medium: Memory
        sizeLimit: 1G

```
2. Check the logs of the pod with: 
```
kubectl -n namespace  logs lorax -f  
```
and the last lines in shows are:
```
2024-03-15T20:10:41.488484Z  INFO lorax_router: router/src/main.rs:465: Serving revision 8cca527612d856d7d32bd94f8103728d614eb852 of model meta-llama/Llama-2-7b-hf
2024-03-15T20:10:41.502080Z  INFO lorax_router: router/src/main.rs:296: Warming up model
```

### Expected behavior

It should start serving the model.

If we look at the GPU stats, it doesn't seem to be using it, and the memory seems on bounds:
```
[0] NVIDIA L4 | 67°C,   0 % | 19048 / 23034 MB |
```

**Comment by tgaddair** [comment]
Hey @martinbomio, do you still observe this effect if you reduce the `max-input-length`, `max-total-tokens`, and `max-batch-prefill-tokens` to something very modest like 512, 1024, and 1024 respectively? Issues during warmup usually indicate the hardware cannot handle the runtime params.

Also worth mentioning that now that we have the `--adapter-memory-fraction` param that defaults to 0.1, you should not need to set aside additional memory using `--cuda-memory-fraction`.

**Comment by martinbomio** [comment]
Hi @tgaddair . Thanks for the quick reply. When using those parameters the server starts up without issues, but those parameters are a lot lower than the default ones (which also work without issues). The default parameters are: `max-input-length= 1792 `, `max-total-tokens= 2048 `, `max-batch-prefill-tokens=4096`

**Comment by rajasbansal** [comment]
@martinbomio try editing the max_position embeddings in the model config directly. Probably related to https://github.com/predibase/lorax/issues/348

**Comment by martinbomio** [comment]
@rajasbansal  I am not exceeding the context-length of the llama2 model. I was actually able to start the server when upgrading my gpu type from L4 to a100. 

@tgaddair  I wonder if there something off in the warmup code, seems like there is when it hangs. 
For another data point, today I ran into the same issue when using the llama2-13b-hf model with a100 and very similar parameters:
```
        - '--model-id=meta-llama/Llama-2-13b-hf'
        - '--source=hub'
        - '--max-input-length=1025'
        - '--max-total-tokens=4096'
        - '--num-shard=1'
```

I can load this models on other inference engines, 7b with l4 and 13b with a100

**Comment by tgaddair** [comment]
Hey @martinbomio, I'm going to take a look at this today and see if I can find a way to reproduce the hanging.

Was the a100 40GB of VRAM or 80GB?

**Comment by tgaddair** [comment]
Hey @martinbomio, I just ran a test on an a100 with 40GB of VRAM using the configuration provided:

```
        - '--model-id=meta-llama/Llama-2-13b-hf'
        - '--source=hub'
        - '--max-input-length=1025'
        - '--max-total-tokens=4096'
        - '--num-shard=1'
```

Everything started fine, but it did take about about 90 seconds for the model to warmup. That's because during warmup we generate `max_total_tokens - max_input_length` tokens to verify that the model is capable of generating that many tokens without error (otherwise it can lead to fatal crashes at runtime).

I noticed you're running on k8s. Something w've observed is that the tqdm progress bars we use to show progress during warmup don't render in the pod logs for k8s. Is it possible that the server was just taking a while to warmup? Could you try launching again and giving a good 3 minutes or so (to err on the cautious side) to warmup?

**Comment by martinbomio** [comment]
we are using 80gb a100 instances.

>I noticed you're running on k8s. Something w've observed is that the tqdm progress bars we use to show progress during warmup don't render in the pod logs for k8s. Is it possible that the server was just taking a while to warmup? Could you try launching again and giving a good 3 minutes or so (to err on the cautious side) to warmup?

yeah I do not see the progress bar. For our deployment, I waited a few hours before terminating it, and it was still in the warm up (or it hanged). Would you mind sharing the whole pod spec (or deployment) you just used?

**Comment by tgaddair** [comment]
Thanks @martinbomio, definitely sounds like a legitimate hang then. Let me try a few things with the pod configuration to see if I can repro. At first glance, I would suspect the issue might be host memory related, as 16GB seems a bit low, but let me try a few things before offering specific recommendations.

**Comment by tgaddair** [comment]
@martinbomio here's what my investigation has turned up so far:

Using your same configuration, when running on a 24GB A5000 GPU (similar to 24GB L4) with the original config, I can repro the hang.

Here is the stack trace of the process:

```
Thread 568 (idle): "MainThread"
    inner (tqdm/utils.py:195)
    fp_write (tqdm/std.py:455)
    print_status (tqdm/std.py:462)
    display (tqdm/std.py:1496)
    refresh (tqdm/std.py:1348)
    update (tqdm/std.py:1243)
    warmup (lorax_server/models/flash_causal_lm.py:790)
    Warmup (lorax_server/server.py:81)
    _unary_interceptor (grpc/_aio_server.py:73)
    intercept (lorax_server/interceptor.py:38)
    invoke_intercept_method (grpc_interceptor/server.py:165)
    _run (asyncio/events.py:80)
    _run_once (asyncio/base_events.py:1909)
    run_forever (asyncio/base_events.py:603)
    run_until_complete (asyncio/base_events.py:636)
    run (asyncio/runners.py:44)
    serve (lorax_server/server.py:324)
    serve (lorax_server/cli.py:89)
    wrapper (typer/main.py:683)
    invoke (click/core.py:782)
    invoke (click/core.py:1434)
    invoke (click/core.py:1688)
    _main (typer/core.py:216)
    main (typer/core.py:778)
    __call__ (click/core.py:1157)
    __call__ (typer/main.py:311)
    <module> (lorax-server:8)
Thread 761 (idle): "Thread-1"
    select (selectors.py:416)
    wait (multiprocessing/connection.py:931)
    wait_result_broken_or_wakeup (concurrent/futures/process.py:385)
    run (concurrent/futures/process.py:320)
    _bootstrap_inner (threading.py:1016)
    _bootstrap (threading.py:973)
Thread 763 (idle): "Thread-2"
    acquire (tqdm/std.py:104)
    __enter__ (tqdm/std.py:111)
    run (tqdm/_monitor.py:66)
    _bootstrap_inner (threading.py:1016)
    _bootstrap (threading.py:973)
Thread 769 (idle): "Thread-3 (_poll_wrapper)"
    run (threading.py:953)
    _bootstrap_inner (threading.py:1016)
    _bootstrap (threading.py:973)
Thread 804 (idle): "Thread-4"
    acquire (tqdm/std.py:104)
    __enter__ (tqdm/std.py:111)
    run (tqdm/_monitor.py:66)
    _bootstrap_inner (threading.py:1016)
    _bootstrap (threading.py:973)
```

The interesting thing here is that it appears to be hanging in tqdm, which is odd, but there do appear to be some reports of containerized systems hanging when there is backpressure on stdout/stderr. Right now I'm seeing if we can either fix or remove the progress bars as a solution. I've confirmed that removing the progress bars in a test branch resolves the issue.

Also, if you set the max input length to be very close to the max total tokens, you shouldn't see the hang, as it only occurs about 3k tokens into warmup.

**Comment by tgaddair** [comment]
@martinbomio @rajasbansal fix is up: #352.

**Comment by martinbomio** [comment]
@tgaddair I can confirm that the issue is fixed and the server is starting up without issues now. Thanks for looking into this!
